subscribe.go

```diff
@@ -85,3 +93,60 @@ func RegisterSubscriber(scheme string, factory SubscriberFactory) {
 	}
 	subReg[scheme] = factory
 }
+
+// ----------------------------------------------------------------------------
+
+// InMemSubscriber is a subscriber, that consumes messages from seeded data.
+// It is useful mainly for testing.
+type InMemSubscriber struct {
+	mu   sync.Mutex
+	msgs map[string][]SubMessage
+}
+
+// NewInMemSubscriber returns new subscriber, that consumes messages from seeded data.
+func NewInMemSubscriber(messagesByTopic map[string][]SubMessage) *InMemSubscriber {
+	return &InMemSubscriber{
+		msgs: messagesByTopic,
+	}
+}
+
+// Subscribe subscribes to in-memory messages by topic.
+func (s *InMemSubscriber) Subscribe(ctx context.Context, topic string, handler Handler) error {
+	for {
+		// check ctx/cancel BEFORE shift-ing each message:
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		msg, ok := s.shiftMessage(topic)
+		if !ok {
+			return nil
+		}
+
+		if err := handler.Handle(msg); err != nil {
+			return err
+		}
+	}
+}
+
+func (s *InMemSubscriber) shiftMessage(topic string) (SubMessage, bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	msgs := s.msgs[topic]
+
+	if len(msgs) == 0 {
+		return nil, false
+	}
+
+	s.msgs[topic] = msgs[1:]
+	return msgs[0], true
+}
+
+// Close forgets any pending messages.
+func (s *InMemSubscriber) Close() error {
+	s.mu.Lock()
+	s.msgs = nil
+	defer s.mu.Unlock()
```

Comment:
strange, if you defer, defer a line above

Reply:
:man_facepalming: bad copypaste